File: answer_dynamics.py
#!/usr/bin/env python3
"""
answer_dynamics.py — Multi-answer chains and unresolved chains analysis.

Outputs:
  - answer_chains.json: distribution of answers per question, chains by length
  - longest_unresolved.json: questions with most back-and-forth but never resolved
  - answer_length_stats.json: answer length vs resolution correlation
"""

import json
from pathlib import Path
from collections import Counter

SCRATCH  = Path("/scratch/b6ar/trvbale.b6ar")
DATA     = SCRATCH / "all_qa.json"
OUT_DIR  = SCRATCH / "embeddings"

def main():
    with open(DATA) as f:
        data = json.load(f)

    episodes = data.get("episodes", [])

    # ── Answer count distribution ─────────────────────────────
    ans_dist    = Counter()   # n_answers -> count
    resolved_by_ans = {}       # n_answers -> {total, resolved}
    total_overturned = 0

    for ep in episodes:
        for q in ep.get("questions", []):
            n = len(q.get("answers", []))
            ans_dist[n] += 1
            if n not in resolved_by_ans:
                resolved_by_ans[n] = {'total': 0, 'resolved': 0}
            resolved_by_ans[n]['total'] += 1
            if q.get('resolved'):
                resolved_by_ans[n]['resolved'] += 1
            total_overturned += sum(1 for a in q.get('answers', []) if a.get('overturned'))

    # ── Longest unresolved chains ─────────────────────────────
    unresolved_chains = []
    for ep in episodes:
        for q in ep.get("questions", []):
            if not q.get('resolved') and len(q.get('answers', [])) >= 4:
                unresolved_chains.append({
                    'episode': ep.get('episode', '?'),
                    'question': (q.get('question') or '')[:120],
                    'n_answers': len(q.get('answers', [])),
                    'answers': [(a.get('answer') or '')[:80] for a in q.get('answers', [])],
                    'topics': q.get('topics', [])[:5],
                })

    unresolved_chains.sort(key=lambda x: -x['n_answers'])

    # ── Answer length stats ────────────────────────────────────
    short_answers  = 0   # <= 50 chars
    med_answers    = 0   # 50-200 chars
    long_answers   = 0   # > 200 chars
    short_resolved = 0
    med_resolved   = 0
    long_resolved  = 0

    for ep in episodes:
        for q in ep.get("questions", []):
            for a in q.get('answers', []):
                txt = a.get('answer') or ''
                l   = len(txt)
                if l <= 50:
                    short_answers += 1
                    if q.get('resolved'): short_resolved += 1
                elif l <= 200:
                    med_answers += 1
                    if q.get('resolved'): med_resolved += 1
                else:
                    long_answers += 1
                    if q.get('resolved'): long_resolved += 1

    # ── Print summary ─────────────────────────────────────────
    print(f"Answer Dynamics Summary:")
    print(f"  Total answers: {sum(ans_dist.values()):,}")
    print(f"  Overturned answers: {total_overturned}")
    print(f"  Questions with most answers: {ans_dist.most_common(1)[0]}")
    print(f"  Longest unresolved chains: {len(unresolved_chains)} with 4+ answers")
    print(f"  Answer length — short (<=50): {short_answers} (res: {short_resolved})")
    print(f"  Answer length — medium (50-200): {med_answers} (res: {med_resolved})")
    print(f"  Answer length — long (>200): {long_answers} (res: {long_resolved})")

    # Save
    chains_out = {
        'answer_distribution': dict(sorted(resolved_by_ans.items())),
        'total_answers': sum(ans_dist.values()),
        'total_overturned': total_overturned,
        'longest_unresolved': unresolved_chains[:50],
        'summary': {
            'pct_multi_answer': round(sum(k*v for k,v in ans_dist.items() if k>=2) / sum(ans_dist.values()) * 100, 1),
            'avg_answers_per_question': round(sum(k*v for k,v in ans_dist.items()) / sum(ans_dist.values()), 2),
        },
        'answer_length_stats': {
            'short':   {'count': short_answers,  'resolved': short_resolved,  'rate': round(short_resolved/short_answers, 3) if short_answers else 0},
            'medium': {'count': med_answers,    'resolved': med_resolved,    'rate': round(med_resolved/med_answers, 3) if med_answers else 0},
            'long':   {'count': long_answers,   'resolved': long_resolved,   'rate': round(long_resolved/long_answers, 3) if long_answers else 0},
        }
    }

    with open(OUT_DIR / "answer_chains.json", "w") as f:
        json.dump(chains_out, f, indent=2)
    print(f"Saved → {OUT_DIR / 'answer_chains.json'}")

if __name__ == '__main__':
    main()

File: dashboard/src/pages/AnswerDynamics.js
/**
 * AnswerDynamics.js — Multi-answer chains and answer length analysis.
 */
import { loadJSON } from '../lib/data.js'

const BASE = ''

function escHtml(s) { return String(s).replace(/&/g,'&amp;').replace(/</g,'&lt;').replace(/>/g,'&gt;') }
function pct(n, d) { return d ? `${(n/d*100).toFixed(1)}%` : '—' }

export async function renderAnswerDynamicsPage(container, store) {
  container.innerHTML = `<div class="loading"><div class="spinner"></div>Loading...</div>`
  try {
    const [chains, james] = await Promise.all([
      loadJSON('answer_chains.json'),
      loadJSON('james_accuracy.json'),
    ])
    renderPage(container, chains, james)
  } catch (e) {
    container.innerHTML = `<div class="loading"><div class="spinner"></div><p style="color:#e74c3c">Could not load answer dynamics data.</p></div>`
  }
}

function renderPage(container, chains, james) {
  const s   = chains.summary
  const als = chains.answer_length_stats

  container.innerHTML = `
    <div class="page-header">
      <h1>🔄 Answer Dynamics</h1>
      <p>How many callers typically weigh in before resolution? What answer length works best?</p>
    </div>

    <!-- Key stats -->
    <div class="stats-grid" style="--cols:4">
      <div class="stat-card"><div class="stat-value">${chains.total_answers.toLocaleString()}</div><div class="stat-label">Total Answers</div></div>
      <div class="stat-card"><div class="stat-value">${s.avg_answers_per_question}</div><div class="stat-label">Avg Answers per Question</div></div>
      <div class="stat-card"><div class="stat-value">${s.pct_multi_answer}%</div><div class="stat-label">Questions with Multiple Answers</div></div>
      <div class="stat-card"><div class="stat-value" style="color:var(--color-red)">${chains.total_overturned}</div><div class="stat-label">Overturned Answers</div></div>
    </div>

    <!-- Answer length vs resolution -->
    <div class="card">
      <h2>📏 Answer Length vs Resolution Rate</h2>
      <div style="display:grid;grid-template-columns:repeat(3,1fr);gap:16px;margin-top:16px">
        ${[['short', 'Short (≤50 chars)'],['medium','Medium (50-200)'],['long','Long (>200)']].map(([key, label]) => {
          const stat = als[key]
          return `<div style="background:var(--color-surface);border-radius:8px;padding:16px;text-align:center">
            <div style="font-size:32px;font-weight:700">${pct(stat.resolved, stat.count)}</div>
            <div style="font-size:14px;font-weight:600;margin-top:4px">${label}</div>
            <div style="font-size:13px;color:var(--color-muted);margin-top:4px">${stat.count.toLocaleString()} answers · ${stat.resolved} resolved</div>
            <div style="margin-top:12px;background:var(--color-primary);height:8px;border-radius:4px;width:${pct(stat.resolved, stat.count)}"></div>
          </div>`
        }).join('')}
      </div>
      <p style="margin-top:16px;font-size:13px;color:var(--color-muted)">Longer answers are strongly associated with higher resolution rates — thorough explanations work.</p>
    </div>

    <!-- Answer count distribution -->
    <div class="card">
      <h2>📊 How Many Answers Before Resolution?</h2>
      <div style="display:flex;flex-wrap:wrap;gap:8px;margin-top:12px">
        ${Object.entries(chains.answer_distribution).sort((a,b)=>Number(a[0])-Number(b[0])).slice(0,8).map(([n, data]) => `
          <div style="background:var(--color-surface);border-radius:8px;padding:12px;text-align:center;min-width:80px">
            <div style="font-size:24px;font-weight:700">${n}</div>
            <div style="font-size:12px;color:var(--color-muted)">answers</div>
            <div style="font-size:13px;margin-top:4px">${data.total.toLocaleString()} Q</div>
            <div style="font-size:12px;color:var(--color-green)">${pct(data.resolved, data.total)} resolved</div>
          </div>
        `).join('')}
      </div>
    </div>

    <!-- Longest unresolved chains -->
    <div class="card">
      <h2>🌀 Longest Unresolved Chains (never got a definitive answer)</h2>
      <div class="unresolved-list">
        ${(chains.longest_unresolved || []).slice(0, 15).map(item => `
          <div style="border-bottom:1px solid var(--color-border);padding:12px 0">
            <div style="font-size:15px;font-weight:600;margin-bottom:4px">${escHtml(item.question)}</div>
            <div style="font-size:13px;color:var(--color-muted)">
              <a href="/episodes?ep=${item.episode}" class="nav-link" data-link>${item.episode}</a>
              · ${item.n_answers} answers · ${(item.topics || []).slice(0,3).join(', ') || 'no topics'}
            </div>
          </div>
        `).join('')}
      </div>
    </div>
  `

  container.querySelectorAll('a[data-link]').forEach(a => {
    a.addEventListener('click', e => { e.preventDefault(); history.pushState(null,'',a.href); window.dispatchEvent(new PopStateEvent('popstate')) })
  })
}